plugins/modeltraining/pkg/modeltraining/aggregation.go

```diff
@@ -0,0 +1,169 @@
+		}
+	}
+	cluster := a.ByCluster[bucket.Key.ClusterID]
+	if _, ok := cluster.ByNamespace[bucket.Key.NamespaceName]; !ok {
+		cluster.ByNamespace[bucket.Key.NamespaceName] = &NamespaceAggregation{
+			ByDeployment: make(map[string]*DeploymentLogCount),
+		}
+	}
+	a.ByCluster[bucket.Key.ClusterID].ByNamespace[bucket.Key.NamespaceName].ByDeployment[bucket.Key.DeploymentName] = &DeploymentLogCount{
+		DeploymentName: bucket.Key.DeploymentName,
+		Count:          bucket.DocCount,
+	}
+}
+
+func (s *ModelTrainingPlugin) run_aggregation() {
+	for {
+		request := map[string]interface{}{
+			"size": 0,
+			"query": map[string]interface{}{
+				"bool": map[string]interface{}{
+					"must": []map[string]interface{}{
+						{
+							"match": map[string]interface{}{
+								"log_type": "workload",
+							},
+						},
+						{
+							"regexp": map[string]interface{}{
+								"kubernetes.namespace_name.keyword": ".+",
+							},
+						},
+						{
+							"regexp": map[string]interface{}{
+								"deployment.keyword": ".+",
+							},
+						},
+					},
+				},
+			},
+			"aggs": map[string]interface{}{
+				"bucket": map[string]interface{}{
+					"composite": map[string]interface{}{
+						"size": 4,
+						"sources": []map[string]interface{}{
+							{
+								"cluster_id": map[string]interface{}{
+									"terms": map[string]interface{}{
+										"field": "cluster_id",
+									},
+								},
+							},
+							{
+								"namespace_name": map[string]interface{}{
+									"terms": map[string]interface{}{
+										"field": "kubernetes.namespace_name.keyword",
+									},
+								},
+							},
+							{
+								"deployment_name": map[string]interface{}{
+									"terms": map[string]interface{}{
+										"field": "deployment.keyword",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		resultAgg := newAggregations()
+		for {
+			var buf bytes.Buffer
+			//var r map[string]interface{}
+			if err := json.NewEncoder(&buf).Encode(request); err != nil {
+				log.Fatalf("Error getting response: %s", err)
+
+			}
+			res, err := s.osClient.Get().Search(
+				s.osClient.Get().Search.WithContext(context.Background()),
+				s.osClient.Get().Search.WithIndex("logs"),
+				s.osClient.Get().Search.WithBody(&buf),
+				s.osClient.Get().Search.WithTrackTotalHits(true),
+				s.osClient.Get().Search.WithPretty(),
+			)
+			if err != nil {
+				log.Fatalf("Error getting response: %s", err)
+			}
+			defer res.Body.Close()
+			var result SearchResponse
+			if err := json.NewDecoder(res.Body).Decode(&result); err != nil {
+				log.Fatalf("Error parsing the response body: %s", err)
+			}
+			for _, b := range result.Aggregations.Bucket.Buckets {
+				resultAgg.Add(b)
+			}
+			after_key := result.Aggregations.Bucket.AfterKey
+			if after_key != nil {
+				((request["aggs"].(map[string]interface{})["bucket"]).(map[string]interface{})["composite"]).(map[string]interface{})["after"] = after_key
```

Comment:
Refactor this to read from the result struct

Reply:
An alternative could be to use gjson as a shortcut to maintaining structs - https://github.com/tidwall/gjson

Reply:
In this case I believe the structs are already in use here